api/mfa.go

```diff
@@ -51,3 +58,45 @@ func (a *API) DisableMFA(w http.ResponseWriter, r *http.Request) error {
 	}
 	return sendJSON(w, http.StatusOK, user)
 }
+
+func (a *API) GenerateRecoveryCodes(w http.ResponseWriter, r *http.Request) error {
+	const NUM_RECOVERY_CODES = 8
+	const RECOVERY_CODE_LENGTH = 8
+	ctx := r.Context()
+	user := getUser(ctx)
+	instanceID := getInstanceID(ctx)
+	if !user.MFAEnabled {
+		return MFANotEnabledError
+	}
+	now := time.Now()
+	recoveryCodeModels := []*models.RecoveryCode{}
+	var terr error
+	var recoveryCode string
+	var recoveryCodes []string
+	var recoveryCodeModel *models.RecoveryCode
+	for i := 0; i < NUM_RECOVERY_CODES; i++ {
+		recoveryCode = crypto.SecureToken(RECOVERY_CODE_LENGTH)
+		recoveryCodeModel, terr = models.NewRecoveryCode(user, recoveryCode, &now)
+		if terr != nil {
+			return internalServerError("Error creating recovery code").WithInternalError(terr)
+		}
+		recoveryCodes = append(recoveryCodes, recoveryCode)
+		recoveryCodeModels = append(recoveryCodeModels, recoveryCodeModel)
+	}
+	terr = a.db.Transaction(func(tx *storage.Connection) error {
+		for _, recoveryCodeModel := range recoveryCodeModels {
+			if terr = tx.Create(recoveryCodeModel); terr != nil {
+				return terr
+			}
+		}
+
+		if terr := models.NewAuditLogEntry(tx, instanceID, user, models.GenerateRecoveryCodesAction, r.RemoteAddr, nil); terr != nil {
+			return terr
+		}
+		return nil
+	})
+
```

Comment:
we need to check for the err here in case the transaction fails (e.g. one of the inserts for the recovery codes fail) and return a 500